### ISSUE TITLE
Implement SGR dim?

### ISSUE THREAD

**Issue by maximbaz** [kovidgoyal/kitty#446]
Hello,

I'm migrating my alacritty configuration to kitty, and while configuring colors I noticed that alacritty has a whole separate section for "dim" colors besides "normal" and "bright", while kitty does not allow defining "dim" colors.

How is it working today, is kitty approximating "normal" colors to get "dim" versions of those? If yes, since I have exact colors from my theme, I'd prefer to use those instead of relying on the approximation.

This is the PR that introduced dim colors for alacritty: https://github.com/jwilm/alacritty/pull/633

**Comment by kovidgoyal** [comment]
I'm not sure what you mean. In kitty you can configure all 256 colors that can be accessed via ANSI escape codes. If you mean support for the DIM SGR code then that is a legacy code intended for cathode ray tube based hardware terminals, it doesn't really make sense any more, and kitty does not support it. And as far as I know most terminals either dont support it or are intending to drop it

**Comment by maximbaz** [comment]
I'm not sure in terminology too, in alacritty's PR it was called "VTE 'dim' flag". Judging by the discussion in this PR (https://github.com/jwilm/alacritty/pull/492), this is supported in alacritty, Konsole, urxvt and iTerm2. 

I found one way how to use dim colors in alacritty - via `tput dim` (will show below). In the PR discussion people used [colorama/demo01.py](https://raw.githubusercontent.com/tartley/colorama/master/demos/demo01.py) to showcase the behavior of "dim" colors.

In alacritty, I can configure "normal", "bright" and "dim" colors, and then:

```
$ tput setaf 3; echo 'aa'              # this renders "yellow - normal"
$ tput setaf 11; echo 'aa'             # this renders "yellow - bright"
$ tput dim; tput setaf 11; echo 'aa'   # this renders "yellow - normal"
$ tput dim; tput setaf 3; echo 'aa'    # this renders "yellow - dim"
```

In kitty, I can configure "color3" and "color11", and then:

```
$ tput setaf 3; echo 'aa'              # this renders "yellow - normal"
$ tput setaf 11; echo 'aa'             # this renders "yellow - bright"
$ tput dim; tput setaf 11; echo 'aa'   # this renders "yellow - bright"
$ tput dim; tput setaf 3; echo 'aa'    # this renders "yellow - normal"
```

Basically, kitty ignores this dim flag. Does this not make sense to support? If I use apps that were made with colorama, I will not see as many colors in kitty as I can see in the rest of the terminals.

**Comment by kovidgoyal** [comment]
Yeah that's the SGR dim and no kitty is not going to support it. As I said, it is a holdover from the days of hardware terminals. I'm pretty sure it is slated for removal from VTE as well, although I'm not at my computer so I cant do a search to find the reference for it. There is no point to it in a software terminal that supports full true color and fully theme-able 8-bit colors. If you want dim yellow you can get it by using a direct 24bit color or by using a 8-bit color. Like this:

```
printf '\033[38:5:178m this is dim yellow' (using 8-bit colors)
or
printf '\033[38:2:215:175:0m this is dim yellow' (using 24bit colors)
```

Oh and there is no need to use tput to do any of this. All terminals support essentially the same set of SGR codes for colors. I dont know why people still use tput/terminfo for colors. See https://en.wikipedia.org/wiki/ANSI_escape_code#SGR_(Select_Graphic_Rendition)_parameters
and
https://gist.github.com/XVilka/8346728

**Comment by maximbaz** [comment]
All right, basically my primary concern was the existence of tools that use this old SGR dim instead of normal colors. But so far I haven't seen any actual example of this being an issue, and I'm all up for reducing the amount of unnecessary legacy in the world 🙂

Regarding `tput`, that was just the first result in google on how to use a dim color 😉 

Thanks for the examples and the explanation.

**Comment by jordwalke** [comment]
The wildly popular terminal test runner [Jest](https://github.com/facebook/jest) makes very good use of Dim and I have found many use cases that 256/trucolor cannot easily replace. For example, when you want to write a terminal program for showing errors and dim some portion of the output to draw the user's attention to a specific part of the output. I haven't been able to find any other way besides dim, which is actually very well supported in iTerm. The reason it is challenging is because for command line output, you want to use the current terminal theme, and not invent a new color scheme. You don't know which hue of red will look decent for the user's current background (when they have configured their 8/16 colors). So by using dim, you're saying that the terminal/user can decide what it means to have a subdued version of a color. Imagine if I chose "solarized" as the color scheme for my error output, but the user's terminal was configured with a totally different set of colors. It wouldn't look natural. But with dim, I can create very beautiful output that looks great regardless of what color scheme the user has configured their terminal with.

<img width="734" alt="dim" src="https://user-images.githubusercontent.com/977348/39953561-73ca5556-5563-11e8-9178-3956224365d5.png">

That is the output from a command line program that dims the code around the red error highlighting, and that program works great with *any* terminal 16 color theme. "dim" is the only way I can imagine implementing a command line app like that.

**Comment by kovidgoyal** [comment]
Sure, you can implement such a program without using dim. Use the escape codes to query the current value of the colors in the terminal and then choose your dim shade accordingly. It's a little more work but works for far more use cases than just dimming something. The only downside is that the program cannot output dim colors until it receives a response from the terminal emulator, which should not matter for a test runner, since time to run tests is going to be orders of magnitude higher that RTT to the terminal. 

For reference, you can look up the query codes here: http://invisible-island.net/xterm/ctlseqs/ctlseqs.html

**Comment by kovidgoyal** [comment]
Also, I should add, that the correct way to have more than 16-themeable colors would be to theme all 256 colors in the 8-bit color table (which kitty supports). Then in your example, to access dim red, the terminal program could just use <ESC>[38:5:52m to change to dull red, which would be themed to match whatever color red is. 

See https://en.wikipedia.org/wiki/ANSI_escape_code#8-bit

While this means more work for theme creators (you now have to theme 256 colors instead of 16 colros) it is a one time cost and it allows terminal programs to have 256 themeable colors, instead of just 16 (or 24 with dim).

This has the virtue of being more elegant code wise, more general, and is supported on more terminals out of the box than dim.

**Comment by kovidgoyal** [comment]
kitty supports changing all 256 colors via its conf file and dynaically at runtime. See https://github.com/kovidgoyal/kitty/blob/master/README.asciidoc#how-do-i-change-the-colors-in-a-running-kitty-instance for the latter.

**Comment by jordwalke** [comment]
I've never heard of being able to query the terminal for their concrete 16 colors. It's cool to hear that this is supported, but how well supported is it by various terminals? Part of the goal is to make it so command line app builders can implement something that tends to work in most places.

**Comment by jordwalke** [comment]
What was the specific color querying command you were referring to? I had a hard time finding it in that large document.

**Comment by kovidgoyal** [comment]
As far as I know it is supported by many terminals, probably more than support dim. See the section Operating System Commands in that document. Those codes allow you to both set and query various colors.

**Comment by kovidgoyal** [comment]
Thinking about it a little, there is one way in which implementing dim makes sense. That is, rather than having 8 configurable colors, dim should cause the foreground color to be rendered with increased alpha, so that it blends more into the background. Advantages:

1) This would work with both light-on-dark color schemes and dark-on-light color schemes, without special configuration by the user
2) it would affect all colors, not just the first eight and it would be closer in spirit to the original use of dim, which, I believe was to reduce the intensity of the electron beam in a CRT display, and not to manipulate colors at all.

This would actually work with @jordwalke use case as well, since the dimmed regions of text would fade comapred to the non-dimmed regions.

I think it should be fairly easy to implement in kitty, since kitty already has the machinery to alpha blend foreground colors (it is used for the inactive_text_alpha option).

That said, this is just a proposal, I am ot particularly keen to implement it. Comments are welcome.

@egmontkob thoughts?

**Comment by egmontkob** [comment]
Probably the approach that makes the most sense to me is what you're doing right now: not to support dim at all :-D

This suggested approach also has some minor inconsistencies, e.g. the final look (color) of the U+2588 FULL BLOCK character depends on the background.

That being said, I think it's a good idea, definitely a reasonable one.

**Comment by jordwalke** [comment]
If it's entirely possible for people to implement Dim in "user space" by manually blending the colors in 256 terminals, then that is also a nice approach. No one seems to do anything like that as far as I can tell. I believe that iTerm takes the approach you described for Dim (lowering the alpha) and it works really well. This kind of Dim has one other valuable purpose which is really hard (impossible?) to implement otherwise (happy to be corrected if wrong): You can output colors with dims, then after the colors are output, the user can change their terminal theme and all the colors transfer over into the new output, along with their dimmings. In this screenshot, I output the Dim codes in iTerm, and only after the log had been output, did I change the theme to be light. Notice how the original terminal output could be "rendered" accurately in another color scheme since I didn't output hard-coded 256 colors. I think that without Dim, that would be very hard to accomplish in a way that works in nearly all terminals. This use case is important because the whole reason you might switch to another color scheme is that it is too bright outside to read some log output that is on your screen right now, and so you need a switch to a light colorscheme. Often, rerunning a build to regenerate the output in the new color scheme wouldn't work. With Dim, you can toggle themes. The eight color pallet is actually kind of brilliant in this respect - it's just that it's too limited with only 8 colors. Dim kind of comes to the rescue.

**Comment by kovidgoyal** [comment]
That works automatically with my second approach of themeing 256 colors and using the dimmer variants from them in the application. When you change the theme, the dim colors will be automatically remapped, in all existing output.

There is not really any use case for DIM that cannot be implemented in other nicer, more portable, but slightly harder ways. 

The question is really whether to support DIM for compatibility with software that does not do the right thing. There are plenty of warts with DIM.

1) It works differently in different terminals (applying to only eight colors, applying to all colors, being user configurable, or not, etc.). It does not work at all in many terminals.

2) It has no SGR code to turn it off (instead the code to turn off bold turns off dim as well, which means you cannot combine DIM with BOLD). 

Given those constraints I'm really surprised application developers choose to use it. I presume people aren't aware of the alternative approaches.

**Comment by jordwalke** [comment]
"That works automatically with my second approach of themeing 256 colors and using the dimmer variants from them in the application"

I think Dim still has an advantage because it allows:
1. An application author to write a program that works on most popular terminals.
2. Dimming effect.
3. Without the user having to configure special color schemes.
4. Where the user can toggle color schemes after the output has been logged.

I believe the 256 pallet solution requires custom configuration on the user's part which they're unlikely to do. I could be wrong, but I'm really glad to see you implemented DIM anyways and I look forward to trying it out. Thanks for working on Kitty! It's excellent.

**Comment by GiorgosXou** [comment]
*(Am I missing something?)* The 6f2d63e commit-message seems missleading. It states that:

> kitty now supports the SGR DIM escape code, which makes text fade into
the background.

but in reality, has no effect at all *(unless I haven't configured something right??)*: 
*(it makes DIM look like NORMAL and DIM+BOLD as NORMAL+BOLD)*:

| ✅ Alacritty | ❌ Kitty |
| --- | --- |
|![Image](https://github.com/user-attachments/assets/d9a76c22-58e0-4816-bcbd-7b0cbc82bce7)|<img  alt="kitty_colors_effects" src="https://github.com/user-attachments/assets/dc98a3f5-d0ec-4950-90be-c54ff1ccb957" />|


<details><summary>(Toggle to see: Example that was used)</summary>

```c
/* 
 * Simply `gcc ./tst_kitty_dim_ncurses.c -lncurses`
 * and then run `./a.out`
*/

#include <ncurses.h>

int main()
{	
	initscr();     // Start curses mode 		  
	start_color(); // Start color  mode

  init_pair(1, COLOR_RED,     COLOR_BLACK);
  init_pair(2, COLOR_GREEN,   COLOR_BLACK);
  init_pair(3, COLOR_YELLOW,  COLOR_BLACK);
  init_pair(4, COLOR_BLUE,    COLOR_BLACK);
  init_pair(5, COLOR_CYAN,    COLOR_BLACK);
  init_pair(6, COLOR_MAGENTA, COLOR_BLACK);
  init_pair(7, COLOR_WHITE,   COLOR_BLACK);

  // using those colors
  for (int i=1; i<8; ++i){
	  attron(COLOR_PAIR(i) | A_STANDOUT | A_BOLD);
	  printw("Hello World !!! STANDOUT BOLD  \n");
	  attroff( A_BOLD);
	  printw("Hello World !!! STANDOUT NORMAL\n");
	  attroff(A_STANDOUT);
	  attron( A_BOLD);
	  printw("Hello World !!! BOLD\n");
	  attroff(A_BOLD);
	  attron(A_NORMAL);
	  printw("Hello World !!! NORMAL\n");
	  attroff(A_NORMAL);
	  attron(A_DIM | A_BOLD);
	  printw("Hello World !!! DIM BOLD\n");
	  attroff(A_BOLD);
	  printw("Hello World !!! DIM\n");
	  attroff(A_DIM);
	  attroff(COLOR_PAIR(i));
	}
	refresh(); // Print it on to the real screen 
	getch  (); // Wait for user input 
	endwin (); // End curses mode		  

	return 0;
}
```

</details>

**Comment by GiorgosXou** [comment]
Nearly any NCurses application uses Dim-effect. It's really weird that it's not implemented.

**Comment by kovidgoyal** [comment]
dim affects opacity aka how faded the foreground color looks, not font weight. Ad you cancontrol that effect via dim_opacity in kitty.conf.

**Comment by GiorgosXou** [comment]
@kovidgoyal  You understand that `A_DIM` has no effect in kitty, right? it has the same exact effect as `A_NORMAL`

**Comment by kovidgoyal** [comment]
Sigh, I have no clue what A_DIM is I assume that's an ncurses-ism. The SGR escape code for DIM reduces foreground opacity, as it is supposed to.

<img width="652" height="62" alt="Image" src="https://github.com/user-attachments/assets/9eb2a830-6fb3-44a1-bdd3-4448b10aca67" />

If ncurses isnt working report the issue to ncurses, or better, use an actual sane and well designed TUI library.

**Comment by GiorgosXou** [comment]
> or better, use an actual sane and well designed TUI library.

Or an actual sane and well designed Terminal.

@ThomasDickey

**Comment by kovidgoyal** [comment]
The terminal's job is to interpret escape codes correctly, which kitty is doing. Furthermore the terminfo database correctly has dim set to `\E[2m`:
```
 infocmp xterm-kitty | grep -i dim
	cvvis=\E[?12;25h, dch=\E[%p1%dP, dch1=\E[P, dim=\E[2m,
```

And dont ping maintainers of other projects randomnly. Use whatever that projects desired channel of communication is. Which for xterm/ncurses is email IIRC.

**Comment by kovidgoyal** [comment]
And just for giggles, in case it wasnt already obvious:
```sh
kitty --dump-commands python -c 'import curses; curses.wrapper(lambda stdscr: (stdscr.addstr(0,0,"Dim Text",curses.A_DIM), stdscr.addstr(1,1, "Normal Text", 0), stdscr.getch()))'
```
shows that ncurses is not setting the dim attribute. Extract from dumped commands:
```
screen_set_mode 7 1
screen_cursor_position 1 1
screen_designate_charset 0 66
select_graphic_rendition 0
select_graphic_rendition 39;49
select_graphic_rendition 37
select_graphic_rendition 40
draw Dim Text
screen_cursor_position 2 2
screen_designate_charset 0 66
select_graphic_rendition 0
select_graphic_rendition 39;49
select_graphic_rendition 37
select_graphic_rendition 40
draw Normal Text
```

**Comment by GiorgosXou** [comment]
You realise you called NCurses a not *"sane and well designed TUI library."* right?

I might be wrong but the fact that it doesn't work as expected is probably a Kitty issue and not an NCurses one.

**Comment by kovidgoyal** [comment]
Because that is exactly what it is. And yes you are wrong, as I took the time and trouble to demonstrate very exactly to you. Good luck, I wont be wasting any more of my time on you.
